pallets/proof-of-stake/src/lib.rs

```diff
@@ -0,0 +1,913 @@
+const REWARDS_PRECISION: u32 = 10000;
+
+pub use pallet::*;
+
+pub mod weights;
+pub use weights::WeightInfo;
+
+type AccountIdOf<T> = <T as frame_system::Config>::AccountId;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {}
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+		type ActivationReservesProvider: ActivationReservesProviderTrait<
+			AccountId = Self::AccountId,
+		>;
+		type NativeCurrencyId: Get<TokenId>;
+		type Xyk: XykFunctionsTrait<Self::AccountId>;
+		type PoolPromoteApi: ComputeIssuance + PoolPromoteApi;
+		type Currency: MultiTokenCurrencyExtended<Self::AccountId>
+			+ MultiTokenReservableCurrency<Self::AccountId>;
+		#[pallet::constant]
+		/// The account id that holds the liquidity mining issuance
+		type LiquidityMiningIssuanceVault: Get<Self::AccountId>;
+		#[pallet::constant]
+		type RewardsDistributionPeriod: Get<u32>;
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::error]
+	/// Errors
+	pub enum Error<T> {
+		/// Not enought assets
+		NotEnoughAssets,
+		/// Division by zero
+		DivisionByZero,
+		/// Math overflow
+		MathOverflow,
+		/// Not enough rewards earned
+		NotEnoughRewardsEarned,
+		/// Not a promoted pool
+		NotAPromotedPool,
+		/// Past time calculation
+		PastTimeCalculation,
+		LiquidityCheckpointMathError,
+		CalculateRewardsMathError,
+		CalculateCumulativeWorkMaxRatioMathError,
+		CalculateRewardsAllMathError,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		PoolPromotionUpdated(TokenId, Option<u8>),
+		LiquidityActivated(T::AccountId, TokenId, Balance),
+		LiquidityDeactivated(T::AccountId, TokenId, Balance),
+		RewardsClaimed(T::AccountId, TokenId, Balance),
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn get_rewards_info)]
+	pub type RewardsInfo<T: Config> = StorageDoubleMap<
+		_,
+		Twox64Concat,
+		AccountIdOf<T>,
+		Twox64Concat,
+		TokenId,
+		RewardInfo,
+		ValueQuery,
+	>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn liquidity_mining_active_pool_v2)]
+	pub type LiquidityMiningActivePoolV2<T: Config> =
+		StorageMap<_, Twox64Concat, TokenId, u128, ValueQuery>;
+
+	// XYK extrinsics.
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::call_index(0)]
+		#[pallet::weight(<<T as Config>::WeightInfo>::proof_of_stake_compound_rewards())]
+		#[transactional]
+		pub fn compound_rewards(
+			origin: OriginFor<T>,
+			liquidity_asset_id: TokenId,
+			amount_permille: Permill,
+		) -> DispatchResultWithPostInfo {
+			let sender = ensure_signed(origin)?;
+
+			<T::Xyk as XykFunctionsTrait<T::AccountId>>::do_compound_rewards(
+				sender,
+				liquidity_asset_id.into(),
+				amount_permille,
+			)?;
```

Comment:
that seems more like ProofOfStakeRewards functionality than XYK to me :thinking:

Reply:
Implementation wise we would still prefer do_compound_rewards to be in Xyk itself.